bigframes/core/rewrite.py

```diff
@@ -196,28 +203,36 @@ def expand(self) -> nodes.BigFrameNode:
         return nodes.ProjectionNode(child=root, assignments=self.columns)
 
 
-def maybe_squash_projection(node: nodes.BigFrameNode) -> nodes.BigFrameNode:
-    if isinstance(node, nodes.ProjectionNode) and isinstance(
-        node.child, nodes.ProjectionNode
-    ):
-        # Conservative approach, only squash consecutive projections, even though could also squash filters, reorderings
-        return SquashedSelect.from_node(node, projections_only=True).expand()
-    return node
-
-
 def maybe_rewrite_join(join_node: nodes.JoinNode) -> nodes.BigFrameNode:
-    left_side = SquashedSelect.from_node(join_node.left_child)
-    right_side = SquashedSelect.from_node(join_node.right_child)
+    rewrite_common_node = common_subtree(join_node.left_child, join_node.right_child)
+    if rewrite_common_node is None:
+        return join_node
+    left_side = SquashedSelect.from_node(join_node.left_child, rewrite_common_node)
+    right_side = SquashedSelect.from_node(join_node.right_child, rewrite_common_node)
     if left_side.can_join(right_side, join_node.join):
```

Comment:
maybe rename it as `can_merge`, I thought it was checking if two trees are joinable or not.

Reply:
done